src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +148,297 @@ public interface IFileWatchingToken
+                => new(Kind.WatchFile, path, fileChangeFlags, sink, token);
+
+            public static WatcherOperation UnwatchDirectories(List<uint> cookies)
+                => new(Kind.UnwatchDirectories, cookies);
+
+            public static WatcherOperation UnwatchFiles(IEnumerable<Context.FileWatchingToken> tokens)
+                => new(Kind.UnwatchFiles, tokens);
+
+            public static WatcherOperation UnwatchFile(Context.FileWatchingToken token)
+                => new(Kind.UnwatchFile, token);
+
+            /// <summary>
+            /// Attempts to combine the current <see cref="WatcherOperation"/> with the next operation in sequence. When
+            /// successful, <paramref name="combined"/> is assigned a value which, when applied, performs an operation
+            /// equivalent to performing the current instance immediately followed by <paramref name="other"/>.
+            /// </summary>
+            /// <param name="other">The next operation to apply.</param>
+            /// <param name="combined">An operation representing the combined application of the current instance and
+            /// <paramref name="other"/>, in that order; otherwise, <see cref="Empty"/> if the current operation cannot
+            /// be combined with <paramref name="other"/>.</param>
+            /// <returns><see langword="true"/> if the current operation can be combined with <paramref name="other"/>;
+            /// otherwise, <see langword="false"/>.</returns>
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                switch (_kind)
+                {
+                    case Kind.WatchDirectory:
+                    case Kind.WatchFile:
+                        // Watching operations cannot be combined
+                        break;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
+                        return true;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
+                        return true;
+
+                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
+                        var cookies = new List<uint>(_cookies);
+                        cookies.AddRange(other._cookies);
+                        combined = UnwatchDirectories(cookies);
```

Comment:
kinda odd (to me) that some are lists, and some are immutable lsits.

Reply:
It's required for correctness (and documented)

Reply:
k

Reply:
I get that the "watch directories" operation takes the mutable list because it's reaching back into the context-held state of the directory watches, but why does the unwatch take a List() versus just an IEnumerable? It seems like that could work as well as UnwatchFile since it's going to be roughly the same pattern?

Reply:
➡️ `FileChangeWatcher.Context` can be disposed prior to the `Watch` operation completing (where completion of the `Watch` operation is the earliest point where the contents of the list are valid).